### PR TITLE
Fixed issue #17776: New security settings

### PR DESCRIPTION
**Pull request by gabrieljenik**

### CODE REVIEW

**Comment by olleharstedt** [comment]
New feature, should be in dev, or?

**Comment by gabrieljenik** [comment]
Well, it is a security issue.
That's why I set it on master.

**Comment by olleharstedt** [comment]
:thinking:

File: application/config/config-defaults.php [hunk 1]
```diff
@@ -804,6 +804,8 @@
 
 $config['pluginWhitelist'] = [];
 
+$config['disablePluginUpload'] = false;
```
Comment (Shnoulle) [inline application/config/config-defaults.php:807]: Add comment please :)

File: application/config/config-defaults.php [hunk 2]
```diff
@@ -804,6 +804,8 @@
 
 $config['pluginWhitelist'] = [];
```
Comment (Shnoulle) [inline application/config/config-defaults.php:805]: Can you add comment on this one ? For https://github.com/LimeSurvey/LimeSurvey/blob/64fb578c5f0fddc0ea89fe4ede27cd44a3f29d82/application/config/config-defaults.php#L805 something like 

> User plugin list allowed to be installed. This don't disallow core plugin, but disable already installed plugin action.

**Review by Shnoulle — CHANGES_REQUESTED** [review]
Need comments

Else : App()->getConfig('pluginCoreList') : really ? In my opibnion : it must be a fixed static variables here

**Comment by olleharstedt** [comment]
For the failures in CI, you can merge master into this branch again. Should be fixed.

File: application/libraries/PluginManager/PluginManager.php [hunk 3]
```diff
@@ -634,17 +638,38 @@ protected function getPluginName(string $class, \ExtensionConfig $extensionConfi
...
+            'AuditLog',
+            'ExportR',
+            'ExportSTATAxml',
+            'ExportSPSSsav',
+            'extendedStartPage',
+            'oldUrlCompat',
+            'AuthLDAP',
+            'Authdb',
+            'Authwebserver'
+        ];
```
Comment (Shnoulle) [inline application/libraries/PluginManager/PluginManager.php:673]: Lack of some core plugin : different ways to fix 

- don't allow some core plugins (maybe disallow Authwebserver for example)
- fix the list adding all core plugins
- get the list from directory (and cache it)

File: application/views/admin/super/fullpagebar_view.php [hunk 4]
```diff
@@ -17,7 +17,8 @@
                 <!-- Install Plugin Zip -->
                 <?php if (
                     isset($fullpagebar['pluginManager']['buttons']['installPluginZipModal']['hasConfigDemoMode']) &&
-                    !$fullpagebar['pluginManager']['buttons']['installPluginZipModal']['hasConfigDemoMode']
+                    !$fullpagebar['pluginManager']['buttons']['installPluginZipModal']['hasConfigDemoMode'] &&
+                    !Yii::app()->getConfig('disablePluginUpload')
```
Comment (Shnoulle) [inline application/views/admin/super/fullpagebar_view.php:21]: Adding a boolean `shownUpload` in controller maybe ? 

PS : i really think this isset is directy …

**Review by Shnoulle — CHANGES_REQUESTED** [review]
More remark than change needed.

But need to be sure for list of core plugins

File: application/libraries/PluginManager/PluginManager.php (hunk 3 above)
Comment (Shnoulle) [reply application/libraries/PluginManager/PluginManager.php:673]: Sorry for the list i give … (copy pasted from yours)

Comment (gabrieljenik) [reply application/libraries/PluginManager/PluginManager.php:673]: Sorry, what do you mean by `Lack of some core plugin : different ways to fix` ?

File: application/views/admin/super/fullpagebar_view.php (hunk 4 above)
Comment (gabrieljenik) [reply application/views/admin/super/fullpagebar_view.php:21]: You mean for not showing the button at all if uploads are disabled?

File: application/libraries/PluginManager/PluginManager.php (hunk 3 above)
Comment (Shnoulle) [reply application/libraries/PluginManager/PluginManager.php:673]: Lack of some plugins : https://github.com/LimeSurvey/LimeSurvey/tree/master/application/core/plugins 
All core plugins not included here :)

Comment (Shnoulle) [reply application/libraries/PluginManager/PluginManager.php:673]: Then the 3 point is 3 way to fix it :) Why allow AuthWebsrevr (that can broke access) and disallow expressionFixedDbVar ?

Comment (Shnoulle) [reply application/libraries/PluginManager/PluginManager.php:673]: In fact your 1st idea to have a config array for this is not a bad solution … but name it 'allowedCorePlugin' and always force AuthDB (currently)

Comment (gabrieljenik) [reply application/libraries/PluginManager/PluginManager.php:673]: I am not sure about changing the name.
If this list is `allowedCorePlugins`, would there be a `notAllowedCorePlugins`? would there be any core plugin not allowed?

Comment (gabrieljenik) [reply application/libraries/PluginManager/PluginManager.php:673]: I suggest that disabling core plugins should be another story.

Comment (Shnoulle) [reply application/libraries/PluginManager/PluginManager.php:673]: > I suggest that disabling core plugins should be another story.

Then you need to fix the list : expressionFixedDbVar, expressionQuestionHelp … etc … :).

By «fix the list adding all core plugins» or «get the list from directory (and cache it)»

Comment (gabrieljenik) [reply application/libraries/PluginManager/PluginManager.php:673]: > fix the list adding all core plugin

Agree.

File: application/views/admin/super/fullpagebar_view.php (hunk 4 above)
Comment (Shnoulle) [reply application/views/admin/super/fullpagebar_view.php]: Thank you

**Comment by Shnoulle** [comment]
Seems action can be done in demoMode ? : another issue or not ?

**Comment by gabrieljenik** [comment]
> Seems action can be done in demoMode ? : another issue or not ?

I think that's another issue.

**Review by Shnoulle — APPROVED** [review]

**Comment by Shnoulle** [comment]
> > Seems action can be done in demoMode ? : another issue or not ?
> 
> I think that's another issue.

Yes : same, you report it after this one is merged ?

**Pull request merged**